lnwallet/musession/manager.go

```diff
@@ -0,0 +1,308 @@
+package musession
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr/musig2"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+// MuSig2State is a struct that holds on to the internal signing session state
+// of a MuSig2 session.
+type MuSig2State struct {
+	// MuSig2SessionInfo is the associated meta information of the signing
+	// session.
+	input.MuSig2SessionInfo
+
+	// context is the signing context responsible for keeping track of the
+	// public keys involved in the signing process.
+	context *musig2.Context
+
+	// session is the signing session responsible for keeping track of the
+	// nonces and partial signatures involved in the signing process.
+	session *musig2.Session
+}
+
+// PrivKeyFetcher is used to fetch a private key that matches a given key desc.
+type PrivKeyFetcher func(*keychain.KeyDescriptor) (*btcec.PrivateKey, error)
+
+// Manager houses the state needed to manage concurrent musig sessions. Each
+// session is identified by a unique session ID which is used by callers to
+// interact with a given session.
+type Manager struct {
+	sync.Mutex
```

Comment:
nit: do we expect users of the Manager to lock it or is this just for the Manager's use? if the later then perhaps consider naming the variable to force it to be unexported

Reply:
So the manager is shared for the entire daemon. This was the state that was previously in the wallet in `lnwallet`. The mutex here is needed to ensure that all access to the set set of shared sessions is thread safe. 

One thing we can do here is have a sub-mutex, for each session, this way we can break things up a bit, and not always require a single global mutex. 

We have the `multimutex` package that can be used for this. Right now we have a version of `unit64` and one for the hash that can actually be consolidated using type params. I can spin that out into a PR first.

Reply:
PR here: https://github.com/lightningnetwork/lnd/pull/7742

Will also cherry-pick that commit here as well. To improve the existing situation.